server.go

```diff
@@ -83,20 +83,53 @@ func checkSameOrigin(r *http.Request) bool {
 	return equalASCIIFold(u.Host, r.Host)
 }
 
-func (u *Upgrader) selectSubprotocol(r *http.Request, responseHeader http.Header) string {
-	if u.Subprotocols != nil {
-		clientProtocols := Subprotocols(r)
-		for _, serverProtocol := range u.Subprotocols {
-			for _, clientProtocol := range clientProtocols {
-				if clientProtocol == serverProtocol {
-					return clientProtocol
-				}
+// firstMatching returns the first matching element present in both slices and true/false whether a match has been found.
+func firstMatching(as []string, bs []string) (string, bool) {
+	for _, a := range as {
+		for _, b := range bs {
+			if a == b {
+				return a, true
 			}
 		}
-	} else if responseHeader != nil {
-		return responseHeader.Get("Sec-Websocket-Protocol")
 	}
-	return ""
+	return "", false
+}
+
+// Subprotocols returns the subprotocols requested by the client in the
+// Sec-WebSocket-Protocol header.
+func Subprotocols(r *http.Request) []string {
+	h := strings.TrimSpace(r.Header.Get("Sec-WebSocket-Protocol"))
+	if h == "" {
+		return nil
+	}
+
+	protocols := strings.Split(h, ",")
+	for i := range protocols {
+		protocols[i] = strings.TrimSpace(protocols[i])
+	}
+	return protocols
+}
+
+// selectSubprotocol returns the first matching subprotocol found, in the following way:
+// -	if Subprotocols in the Upgrader struct is unset and the client's subprotocol is unset (or empty),
+//		it returns ""
+// -	if Subprotocols in the Upgrader struct is set and responseHeader is unset,
+//		it returns the first matching subprotocol from Subprotocols and the r *http.Request
+// -	if responseHeader is set, it returns the first matching subprotocol from the ResponseHeader (ignoring Subprotocols)
```

Comment:
If the application specifies the negotiated subprotocol, then that subprotocol should be used as is.   Because the response header is a single value, it does not make sense to process it as a list.